### ISSUE TITLE
Desktop files with NotShowIn=LXQt are still displayed in the menu

### ISSUE THREAD

**Issue by monetaryabyss** [lxqt/lxqt#1743]
##### Expected Behavior
`.desktop` files in `/usr/share/applications/` show in the LXQt menu however adding `NotShowIn=LXQt;` should hide the desktop item from the menu according to the freedesktop specification, yet still allow it to be seen in other desktop environments.

##### Current Behavior
After adding `NotShowIn=LXQt;` the item is still displayed in the menu even after a logout/login, session restart, or reboot.

##### Steps to Reproduce (for bugs)

1. Create a desktop file in `/usr/share/applications/`or modify an existing one such as this:
```
[Desktop Entry]
Version=1.0
Name=AlsaMixer
Type=Application
Comment=Show ALSA Audio Mixer
Terminal=true
Exec=alsamixer
Icon=multimedia-volume-control
Categories=AudioVideo;Audio;Mixer;
Keywords=Settings;Audio;
NotShowIn=LXQt;
```
2. restart computer
3. find `AlsaMixer` in sound and video menu

##### Context
Some menu items may not wish to be visible in all available desktop environments available to the system.
##### System Information

* Distribution & Version: Lubuntu 19.04
* Kernel: 5.0.0-27 generic
* Qt Version: 5.12.2
* liblxqt Version: 0.14.1
* Package version: 0.14.1

Also tested on

* Distribution & Version: Debian 10 
* Kernel: 4.19.0-5-amd64
* Qt Version: 5.11.3
* liblxqt Version: 0.14.1
* Package version: 0.14.1

**Comment by agaida** [comment]
Right, not implemented i guess - Workaround: OnlyShowIn=None

**Comment by agaida** [comment]
@monetaryabyss  - some remarks:
* forget `OnlyShowIn=None` - the way a user should do this is:
```
cp /usr/share/applications/$foo.desktop ~/.local/share/applications
echo Hidden=true >> ~/.local/share/applications/$foo.desktop
```
Done
* Never ever, i repeat, never ever edit files in /usr/share or other locations who are not configuration directories - these changes will be gone after the next update
* read the freedesktop spezification about Hidden - NotShowIn should be deprecated, negative lists are not the right way to do these things right imho - if a Application should be shown in a certain Environment --> OnlyShowIn, if in more Environments ShowIn=foo;bar;baz - NotShowIn isn't used so often - so we forget to implement :)
* cp and echo (or $EDITOR) is a really fine menu editor - alacarte might work too

**Comment by wxl** [comment]
> read the freedesktop spezification about Hidden - NotShowIn should be deprecated

@agaida Looking at the [Deprecated Items](https://specifications.freedesktop.org/desktop-entry-spec/latest/apc.html) part of the specification, it seems you are actually wrong.

P.S. @monetaryabyss is asking this as a distro developer and less as a user, so editing global files is totally appropriate here.

**Comment by tsujan** [comment]
> @agaida Looking at the Deprecated Items part of the specification, it seems you are actually wrong.

You know that he'll never read it; don't you?

**Comment by wxl** [comment]
Yeah but I was hoping you would instead :)

**Comment by tsujan** [comment]
I have 2 DEs besides LXQt and confirm that the lack of `NotShowIn=LXQt` is a problem (`NoDisplay=true` "solved" my problem because my main DE is LXQt). I always wanted to fix it but never found the time and don't know when I do (currently being busy with Qt6).

**Comment by luis-pereira** [comment]
```diff
++mXdgMenu.setEnvironments(QStringList() << QStringLiteral("X-LXQT") << QStringLiteral("LXQt"));  
--mXdgMenu.setEnvironments(QStringList() << QStringLiteral("LXQt"));
```

at https://github.com/lxqt/lxqt-panel/blob/18acf9efc87a05c467d7c353e8db815de84304dd/plugin-mainmenu/lxqtmainmenu.cpp#L237
solves the issue for LXQt(only).

Why do we add `X-LXQT` to the environment list ? An artifact from when LXQt wasn't a registered  environment ?
What are the side effects of removing `X-LXQT`, if any ?

Of course a general solution is needed.

**Comment by tsujan** [comment]
Somehow related to https://github.com/lxqt/lxqt/issues/1845?

**Comment by wxl** [comment]
@luis-pereira good catch but I think you mean

```diff
--mXdgMenu.setEnvironments(QStringList() << QStringLiteral("X-LXQT") << QStringLiteral("LXQt"));  
++mXdgMenu.setEnvironments(QStringList() << QStringLiteral("LXQt"));
```

It looks like using [X-LXQT comes from the Razor days](https://github.com/lxqt/lxqt-panel/commit/782d446f246ec2cbbe5f04b0fe6ccdb0d2860a91#diff-f7ab3ac13c3dc2cdd714dacdd5fa2ecddd7b51c5540eab6f00f73b1509ebede4):

```diff
-    mXdgMenu.setEnvironments(QStringList() << "X-RAZOR" << "Razor");
+    mXdgMenu.setEnvironments(QStringList() << "X-LXQT" << "LxQt");
```

#1845 seems to be about categories not environments, no?

**Comment by tsujan** [comment]
> #1845 seems to be about categories not environments

Yes.

>  Furthermore, it seems to be related to outdated software.

No! The comment that says so is incorrect. It happens with the latest version of uses desktop-file-utils too.

**Comment by wxl** [comment]
So LXQt is included in the list of [`OnlyShowIn` environments](https://specifications.freedesktop.org/menu-spec/latest/apb.html) but it is *NOT* included in the [Additional Categories](https://specifications.freedesktop.org/menu-spec/latest/apas02.html) where you find other environments like `KDE`, `GNOME`, and `Qt`. If you look at the [examples](https://specifications.freedesktop.org/menu-spec/latest/ar01s03.html) you see `Categories=GNOME;Utility;` which leads you to believe that any of the environments can be used as a category, but that's not the case.

tl;dr LXQt needs to be added to the additional categories to resolve the need to have any `X-`anything.

**Comment by wxl** [comment]
The one thing I don't understand is that we are creating a list of environments right now, one of which is `X-LXQt` and one of which is `LXQt`. The former is valid in `OnlyShowIn` since it's an extension. The latter is valid because it's included in the `OnlyShowIn` environments. And yet, it seems to not work. Maybe there's a problem with using a list?

**Comment by tsujan** [comment]
> ... but it is NOT included in the Additional Categories

Correct. Only KDE, GNOME and XFCE are included, maybe because they're very old. MO, that's acceptable; KDE, GNOME and XFCE can be seen as exceptions to the rule. There are several other DEs — I think Enlightenment was started before GNOME — why should all of them be included when an `X-` solves the "problem"?

**Comment by wxl** [comment]
Yeah I mean I think it's a really silly issue. I would never want to categorize my stuff by library. The only applications that would apply are LXQt ones. That's such a small number, I'm not sure why anyone would care. To be frank, I'm surprised XFCE is even in there.

**Comment by luis-pereira** [comment]
> @luis-pereira good catch but I think you mean
> 
> ```diff
> --mXdgMenu.setEnvironments(QStringList() << QStringLiteral("X-LXQT") << QStringLiteral("LXQt"));  
> ++mXdgMenu.setEnvironments(QStringList() << QStringLiteral("LXQt"));
> ```

Of course.

**Comment by luis-pereira** [comment]
> And yet, it seems to not work. Maybe there's a problem with using a list?

Yes, there is a bug.

**Comment by wxl** [comment]
The question is whether or not the bug is with us or XDG, no? If the list consists of GNOME and KDE (I pick these because they're in both the environments and the categories), do they both work as expected?

**Comment by luis-pereira** [comment]
@wxl There is bug in `libqtxdg`.

**Comment by luis-pereira** [comment]
Closed by https://github.com/lxqt/libqtxdg/pull/328.
